wp_get_font_dir():  Bail if the 'font_dir' filter is already running.

matiasbenedetto opened the pull request:
## What?
`wp_get_font_dir()`: Bail if the `font_dir` filter is already running.

Props to @costdev for [proposing](https://wordpress.slack.com/archives/C02RQBWTW/p1709082902708339?thread_ts=1708987810.984969&cid=C02RQBWTW) this solution.


## Why?
This avoids an infinite loop that can occur if wp_upload_dir() is called inside a 'font_dir' callback.

## Testing instructions
1. Use `wp_get_upload_dir()` inside a function used to filter `upload_dir` as in the following example:

```php
 function alter_wp_fonts_dir( $defaults ) {
	$wp_upload_dir = wp_get_upload_dir();
	$uploads_basedir = $wp_upload_dir['basedir'];
	$uploads_baseurl = $wp_upload_dir['baseurl'];

	$fonts_dir = $uploads_basedir . '/fonts';
	// Generate the URL for the fonts directory from the font dir.
	$fonts_url = str_replace( $uploads_basedir, $uploads_baseurl, $fonts_dir );

	$defaults['path'] = $fonts_dir;
	$defaults['url']  = $fonts_url;

	return $defaults;
}
add_filter( 'font_dir', 'alter_wp_fonts_dir' );
```
2. Upload fonts using the font library
3. Check that the fonsts were successfully uploaded to the path set by the filter.
(`wp-content/uploads/fonts` in the example provided).

---

Trac ticket: https://core.trac.wordpress.org/ticket/60652

matiasbenedetto committed:
wp_get_font_dir():  Bail if the 'font_dir' filter is already running.

This avoids an infinite loop that can occur if wp_upload_dir() is called inside a 'font_dir' callback.

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

matiasbenedetto committed:
format php

swissspidy commented:
The one thing I don't like about `wp_get_upload_dir()` is that it seems like a function that simply returns a directory name. But no, it is actually used for filtering the upload dir.

Things would be much easier if it would behave more like e.g. `wp_privacy_exports_dir()`, or if we split this up:

- 1 function to filter the uploads dir
- 1 function `wp_get_upload_dir()` that returns the desired location and itself has a filter that hosts/devs can use

costdev commented on src/wp-includes/fonts.php:
```diff
@@ -140,7 +150,7 @@ function wp_get_font_dir( $defaults = array() ) {
 	 * @param array $defaults The original fonts directory data.
 	 */
-	return apply_filters( 'font_dir', $defaults );
+	return doing_filter( 'font_dir' ) ? $defaults : apply_filters( 'font_dir', $defaults );
```
Hmm 🤔, only one of the `doing_filter( 'font_dir' )` should be needed. Let's evaluate.

If `doing_filter( 'font_dir' )`, then `wp_get_font_dir()` will have already made its changes.

- Bailing early should be safe.
- Bailing late would overwrite the previously filtered `$defaults`.

@hellofromtonya @matiasbenedetto Can I get a confidence check on that?

costdev replied:
Nevermind, [the alternative approach](https://github.com/WordPress/gutenberg/pull/58839).

costdev commented on src/wp-includes/fonts.php:
```diff
@@ -118,6 +118,16 @@ function wp_unregister_font_collection( string $slug ) {
  * }
  */
 function wp_get_font_dir( $defaults = array() ) {
+	/*
+	* Bail if the 'font_dir' filter is already running.
+	*
+	* This avoids an infinite loop that can occur
+	* if wp_upload_dir() is called inside a 'font_dir' callback.
+	*/
+	if ( doing_filter( 'font_dir' ) ) {
```
```suggestion
	/*
	 * Bail if the 'font_dir' filter is already running.
	 *
	 * This avoids an infinite loop that can occur
	 * if wp_upload_dir() is called inside a 'font_dir' callback.
	 */
	if ( doing_filter( 'font_dir' ) ) {
```
When will IDEs correctly outdent multiline comments?! :joy: This drives me nuts when it happens to me!

matiasbenedetto commented:
> Things would be much easier if it would behave more like e.g. wp_privacy_exports_dir(), or if we split this up:
> 1 function to filter the uploads dir
> 1 function wp_get_upload_dir() that returns the desired location and itself has a filter that hosts/devs can use

@swissspidy I' don't see how `wp_privacy_exports_dir()` is different to the trunk version of `wp_get_font_dir()`. They seem to be doing exactly the same: setting a default and applying a filter after that. Here are the 2 implementations to compare. I removed the comments to make them shorter:

```php
function wp_privacy_exports_dir() {
	$upload_dir  = wp_upload_dir();
	$exports_dir = trailingslashit( $upload_dir['basedir'] ) . 'wp-personal-data-exports/';
	return apply_filters( 'wp_privacy_exports_dir', $exports_dir );
}
```

```php
function wp_get_font_dir( $defaults = array() ) {
	$site_path = '';
	if ( is_multisite() && ! ( is_main_network() && is_main_site() ) ) {
		$site_path = '/sites/' . get_current_blog_id();
	}

	// Sets the defaults.
	$defaults['path']    = path_join( WP_CONTENT_DIR, 'fonts' ) . $site_path;
	$defaults['url']     = untrailingslashit( content_url( 'fonts' ) ) . $site_path;
	$defaults['subdir']  = '';
	$defaults['basedir'] = path_join( WP_CONTENT_DIR, 'fonts' ) . $site_path;
	$defaults['baseurl'] = untrailingslashit( content_url( 'fonts' ) ) . $site_path;
	$defaults['error']   = false;
	
	return apply_filters( 'font_dir', $defaults );
}
```
I think `wp_privacy_exports_dir()` is not entering an infinite loop because it's not used to filter `upload_dir` as the font library does. Otherwise, the same problem would arise.

https://github.com/WordPress/wordpress-develop/blob/a69052a75fa9bfdbf2e4e80a52b39b1c1233ab91/src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php#L859

@swissspidy @costdev What do you think about this alternative  https://github.com/WordPress/gutenberg/pull/58839 to avoid the infinite loop this PR is trying to fix?

matiasbenedetto committed:
fix comment formatting

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

swissspidy commented:
> What do you think about this alternative https://github.com/WordPress/gutenberg/pull/58839 to avoid the infinite loop this PR is trying to fix?

This is exactly what I suggested in my comment :)

> I' don't see how wp_privacy_exports_dir() is different to the trunk version of wp_get_font_dir().
> I think wp_privacy_exports_dir() is not entering an infinite loop because it's not used to filter upload_dir as the font library does.

That's exactly what I said is confusing :)

costdev commented:
Yeah, the alternative looks better to me, nice!

matiasbenedetto commented:
@swissspidy @costdev, if the alternative approach is looking better, could you please add a review in [that PR](https://github.com/WordPress/gutenberg/pull/58839)?

matiasbenedetto commented:
Closing this PR in favor of: https://github.com/WordPress/wordpress-develop/pull/6200
